backend/api/models.py

```
@@ -218,25 +218,30 @@ def get_preference_based_videos(cls, user, preferences, limit, offset=0):
+            # Apply exclusion *before* slicing
+            popular_query = cls.get_popular_videos_queryset().exclude(id__in=recommended_ids)
+            additional_videos = popular_query[additional_offset:additional_offset + additional_needed]
             
             return list(additional_videos)
         return cls.get_popular_videos(limit, offset)
     
     @classmethod
-    def get_popular_videos(cls, limit, offset=0):
-        limit = max(1, min(limit, 50))
-        offset = max(0, offset)       
+    def get_popular_videos_queryset(cls):
```

Ruff (0.8.2)

Missing return type annotation for classmethod `get_popular_videos_queryset`

(ANN206)